### PR TITLE
Implement SimSiam

### LINKED ISSUE #54: Implement SimSiam Representation Learning
As proposed by users u/lfotofilter and u/AiDreamer on Reddit.

Add an implementation of the SimSiam Representation Learning framework.
- [Paper](https://arxiv.org/pdf/2011.10566.pdf)

### PR DESCRIPTION
**Pull request by busycalibrating**
WIP implementation for the SimSiam model. This implementation allows you to use any backbone that you'd like. The model trains, but the results need to be validated. I also refactored `import torchvision.transforms as transforms` to `import torchvision.transforms as T` because there were variables with the same name that could cause bugs. Let me know of whatever feedback you may have!

This would close #54. 

TODO:

- [ ] Validate performance/results
- [x] Implement unit tests
- [x] ~~Create a class to prebuild resnet based backbones (like the SimCLR implementation)~~ 
- [x] Tidy up docstrings

### CODE REVIEW

File: lightly/data/collate.py [hunk 1]
```diff
@@ -304,3 +304,12 @@ def __init__(self,
             rr_prob=rr_prob,
             normalize=imagenet_normalize,
         )
+
+
+class SimSiamCollateFunction(SimCLRCollateFunction):
```
Comment (philippmwirth) [inline lightly/data/collate.py:309]: I wouldn't introduce an extra class, if the settings are the exact same as for SimCLR. For example, we didn't write a new class for the MoCo v2 augmentations because they are exactly the same as SimCLR.

Comment (busycalibrating) [reply lightly/data/collate.py:309]: Agreed - would prefer to keep these augmentations with the simclr name, or pull them into a more generic/high level class of base augmentations?

Comment (philippmwirth) [reply lightly/data/collate.py:309]: I'd say keep them with the SimCLR name

File: lightly/loss/symmetrized_loss.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        if output.shape[0] % 4:
+            raise ValueError('Expected output of shape 4*bsz x dim but got '
+                             f'shape {output.shape[0]} x {output.shape[1]}.')
+
+        # device = output.device
+        batch_size, dim = output.shape
+        batch_size = batch_size // 4
+
+        # normalize the output to length 1
+        output = torch.nn.functional.normalize(output, dim=1)
```
Comment (IgorSusmelj) [inline lightly/loss/symmetrized_loss.py:72]: I think we don't need normalization since we use cosine similarity and not L2 distance.

File: lightly/models/simsiam.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    l1 = nn.Sequential(nn.Linear(in_dims, h_dims),
+                       nn.BatchNorm1d(h_dims),
+                       nn.ReLU(inplace=True))
+
+    l2 = nn.Sequential(nn.Linear(h_dims, h_dims),
+                       nn.BatchNorm1d(h_dims),
+                       nn.ReLU(inplace=True))
+
+    l3 = nn.Sequential(nn.Linear(h_dims, out_dims),
+                       nn.BatchNorm1d(out_dims))
```
Comment (IgorSusmelj) [inline lightly/models/simsiam.py:70]: I'm not sure whether we should add here or on line 33 (before linear layer) another ReLU activation. Currently, we have

- linear
- batch_norm
- linear

two linear layers without non-linearity between seem odd to me. 

I would add ReLU at line 33. That way we don't have a sparse output after the projection head.

File: lightly/loss/symmetrized_loss.py (hunk 2 above)
Comment (busycalibrating) [reply lightly/loss/symmetrized_loss.py:72]: Whoops yes you're right - there's some dead code I need to tidy up

File: lightly/models/simsiam.py (hunk 3 above)
Comment (busycalibrating) [reply lightly/models/simsiam.py:70]: That's true it is not typical. However this would match the implementation described in [the paper (pg 3)](https://arxiv.org/pdf/2011.10566.pdf): 

> _Projection MLP_. The projection MLP _(f)_ has BN applied to each fully-connected (fc) layer, including its output fc. Its **output fc has no ReLU**. The hidden fc is 2048 d. This MLP has 3 layers. 
>
> _Prediction MLP._ The prediction MLP _(h)_ has BN applied to its hidden fc layers.   Its output fc does not have BN (ablation in Sec. 4.4) or ReLU. This MLP has 2 layers. The dimension of _h_’s input and output (_z_ and _p_) is d=2048, and _h_’s hidden layer’s dimension is 512, making _h_ a bottleneck structure (ablation in supplement).

This could be worth experimenting with to see the difference in performance?

File: lightly/data/collate.py (hunk 1 above)
Comment (laurenmoos) [reply lightly/data/collate.py:309]: +1 for this advice, almost did this exact same thing (extend SimCLR for BYOL

File: lightly/loss/symmetrized_loss.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
+import torch
+import numpy as np
+
+class SymmetrizedLoss(torch.nn.Module):
+    """Implementation of the Symmetrized Loss.
```
Comment (philippmwirth) [inline lightly/loss/symmetrized_loss.py:5]: I'd suggest to give this loss a more descriptive name (for example `SymmetricNegCosineSimilarityLoss` or `NegCosineSimilarityLoss`).

File: lightly/models/simsiam.py (hunk 3 above)
Comment (IgorSusmelj) [reply lightly/models/simsiam.py:70]: Got this reply from Kaiming:
Dear Igor,

 

Thank you for your interest!

 

The predictor head by default is: Linear->BN->ReLU->Linear. Also note the dimension is 2048->512->2048. And you are right, there is no ReLU between the output of projection head and the input of prediction head. This is because this embedding (output of projection head) is expected to be normalized to the unit sphere, so ReLU is not desired.

 

Regards,

Kaiming

File: lightly/models/simsiam.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+            TODO
+        pred_hidden_dim:
+            TODO
+        out_dim:
+            Dimension of the output (after the projection head).
+
+    """
+
+    def __init__(self,
+                 backbone: nn.Module,
```
Comment (philippmwirth) [inline lightly/models/simsiam.py:96]: Please also rename `backbone` to `features` for consistency

Comment (busycalibrating) [reply lightly/models/simsiam.py:96]: If you're refactoring the models would you be open to renaming the other names to backbone? I think it's more informative than features in this context

Comment (busycalibrating) [reply lightly/models/simsiam.py:96]: On board with keeping it consistent though obviously

Comment (philippmwirth) [reply lightly/models/simsiam.py:96]: Yes, we also like `backbone` better. In that case you can keep it the way it is and we will change the other names.

File: lightly/models/simsiam.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+        # device = output.device
+        batch_size = len(x)
+        batch_size = batch_size // 2
+
+        x1, x2 = x[:batch_size], x[batch_size:]
+
+        emb1, emb2 = self.backbone(x1), self.backbone(x2)
+        emb1, emb2 = emb1.squeeze(), emb2.squeeze()
+        z1, z2 = self.projection_mlp(emb1), self.projection_mlp(emb2)
+        p1, p2 = self.prediction_mlp(z1), self.prediction_mlp(z2)
```
[the remark this answers is not in the document]
Comment (busycalibrating) [reply lightly/models/simsiam.py:156]: I think what he meant is that based on the design architecture, the output of the embedding is already scattered on the unit hypersphere [(paper pg 3-4)](https://arxiv.org/pdf/2011.10566.pdf):

> The blue curve in Figure 2 (middle) shows that with stop-gradient, the std value is near 1/√d. This indicates that the outputs do not collapse, and they are scattered on the unit hypersphere.

Would you agree with my interpretation?

**Comment by busycalibrating** [comment]
Currently waiting for #66 to go through and then I'll adjust my implementation based on that.

File: lightly/models/simsiam.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (philippmwirth) [reply lightly/models/simsiam.py:156]: The description of the plot you are referencing (Figure 2) is:
> Middleplot:  the per-channel std of the **l<sub>2</sub>-normalized output**, plotted as the averaged std over all channels.

So my interpretation is that the **l<sub>2</sub>-normalized** output of the embeddings is scattered on the unit hypersphere (as opposed to collapsing and basically pointing in a single direction). Therefore I would agree with @IgorSusmelj  here. Although it's not clear to me from the paper that the outputs need to be normalized.

File: lightly/models/simsiam.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+        self.proj_hidden_dim = proj_hidden_dim
+        self.pred_hidden_dim = pred_hidden_dim
+        self.out_dim = out_dim
+
+        self.projection_mlp = \
+            _projection_mlp(num_ftrs, proj_hidden_dim, out_dim)
+
+        self.prediction_mlp = \
+            _prediction_mlp(num_ftrs, pred_hidden_dim, out_dim)
+
```
Comment (philippmwirth) [inline lightly/models/simsiam.py:115]: This should be 
```
self.prediction_mlp = \
    _prediction_mlp(out_dim, pred_hidden_dim, out_dim)
```
Because then I can change the `num_ftrs` without breaking the forward pass, right?

File: lightly/models/simsiam.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (IgorSusmelj) [reply lightly/models/simsiam.py:156]: Maybe just try one implementation. Adding normalization should be rather easy with one line of code. We can add tensorboard logging of the std using pytorch-lightning `self.log(..)` and verify whether the results match with the paper.

[the remark this answers is not in the document]
Comment (philippmwirth) [reply lightly/models/simsiam.py:156]: +1 on this, just leave it like it is for the moment.

[the remark this answers is not in the document]
Comment (busycalibrating) [reply lightly/models/simsiam.py:156]: Sounds good!

File: lightly/models/simsiam.py (hunk 7 above)
Comment (busycalibrating) [reply lightly/models/simsiam.py:115]: If anything, would it not be

```Python
self.prediction_mlp = _prediction_mlp(num_ftrs, pred_hidden_dim, num_ftrs)
```

Not totally sure what you mean by changing `num_ftrs` without breaking the forward pass?

Comment (philippmwirth) [reply lightly/models/simsiam.py:115]: Currently, if `num_ftrs != out_dim`, the forward pass breaks because the output of the projection head has a different shape than the input of the prediction head. Therefore, my suggestion was to change the input dimension of the prediction head to the output dimension of the projection head.

Comment (busycalibrating) [reply lightly/models/simsiam.py:115]: Ah gotcha, yes you're right. I still mix up prediction and projection in my head 😀

**Comment by philippmwirth** [comment]
Btw for the validation of the model we can run experiments when the PR is ready in case you don't have the compute :) 

I already did a few quick tests on cifar10 and the linear evaluation accuracy is around 63% after 50 epochs which roughtly corresponds to what we can see in [Figure D.1](https://arxiv.org/pdf/2011.10566.pdf). So it seems we are on the right path.

File: lightly/loss/sym_neg_cos_sim_loss.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+        Args:
+            x:
+                Predictions with shape: n x d.
+            y: 
+                Projections with shape: n x d. Gradients wll not flow through this
+                variable.
+
+        Returns:
+            [type]: [description]
```
Comment (busycalibrating) [inline lightly/loss/sym_neg_cos_sim_loss.py:37]: Fix this

File: lightly/loss/sym_neg_cos_sim_loss.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+                out1:
+                    Output projections of the second set of transformed images.
+
+            Returns:
+                Contrastive Cross Entropy Loss value.
+
+            Raises:
+                ValueError if shape of output is not multiple of batch_size.
+        """
+        z0, p0 = out0
```
Comment (busycalibrating) [inline lightly/loss/sym_neg_cos_sim_loss.py:61]: Should we do a check for whether the forward pass from the `SimSiam` model was in `return_features` mode?

File: lightly/models/simsiam.py [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+        self.projection_mlp = \
+            _projection_mlp(num_ftrs, proj_hidden_dim, out_dim, num_mlp_layers)
+
+        self.prediction_mlp = \
+            _prediction_mlp(out_dim, pred_hidden_dim, out_dim)
+        
+    def forward(self, 
+                x0: torch.Tensor, 
+                x1: torch.Tensor,
```
Comment (busycalibrating) [inline lightly/models/simsiam.py:130]: @philippmwirth I didn't add support for `x1=None` as per the latest simclr/moco models forward pass. Is this something that you want across the board?

**Comment by busycalibrating** [comment]
I rebased my branch and updated my implementation to account for the API change. I think this is ready for the CIFAR-10 validation, and I just need to implement some tests. Should be good to go soon!

File: lightly/models/simsiam.py (hunk 10 above)
Comment (IgorSusmelj) [reply lightly/models/simsiam.py:130]: It would be great to have a more or less standardized interface across the different models.

**Comment by busycalibrating** [comment]
I added some basic tests (mimicking what was implemented for SimCLR), but I think these tests are pretty basic and at some point should test for numeric results with known inputs. But that should probably be its own issue. 

This should be now ready for proper review and validation

File: lightly/loss/sym_neg_cos_sim_loss.py (hunk 9 above)
Comment (philippmwirth) [reply lightly/loss/sym_neg_cos_sim_loss.py:55]: No, I think it's better to be clear about what inputs the function expects and leave it to the user to pass the right arguments.

File: lightly/loss/sym_neg_cos_sim_loss.py [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+        >>> # generate two random transforms of images
+        >>> t0 = transforms(images)
+        >>> t1 = transforms(images)
+        >>>
+        >>> # feed through SimSiam model
+        >>> output = model(t0, t1)  
+        >>>
+        >>> # calculate loss
+        >>> loss = loss_fn(output)
+
```
Comment (philippmwirth) [inline lightly/loss/sym_neg_cos_sim_loss.py:25]: The example is not up-to-date, it should be:
```python
>>> out0, out1 = model(t0, t1)
>>> loss = loss_fn(out0, out1)
```

and you're missing the brackets for the constructor:
```python
>>> loss_fn = SymmetrizedLoss()
```

Comment (busycalibrating) [reply lightly/loss/sym_neg_cos_sim_loss.py:25]: Good catch!

**Comment by IgorSusmelj** [comment]
I just run a test modifying the MoCo tutorial to use SimSiam (last commit: 826524b2e293b63774a434936adf3ed93d90bd9e). Run with the same settings (100 epochs, resnet-18, etc.)
Final accuracy is 77.9% on the test set

image
image
image

**Review by philippmwirth — APPROVED** [review]
